### PR TITLE
NAT traversal with libp2p

### PR DESCRIPTION
**Pull request by deniskamazur**
Our current strategy is to use [go-libp2p-daemon](https://github.com/libp2p/go-libp2p-daemon) to connect between users
* __Compile libp2p daemon__
   * [x] Build libp2p daemon on setup
      * [x] Ensure that we build a specific version of go-libp2p-daemon
   * [x] Make sure p2pd is included with pypi package
   * [x] Find out the optimal parameters for nat traversal

* __Implement a P2Pd controller (codename hivemind.P2P)__
   * [ ] Make sure it always kills the p2pd process on termination (both with del and with kill -9)
   * [x] add_rpc_servicer (simlar to GRPC)
       * [x] Ensure that it supports asyncio
       * [ ] Ensure that it is process-safe
   * [x] Add unit tests for hivemind.P2P
   * [x] Figure out the user experience (e.g. spawn P2P when DHT is created)
   * [x] Figure out the codebase layout (e.g. hivemind.dht.p2p)
      * Considerations:
          * Move synchronization and get_dht_time to P2P? (rationale: support clock correction)
          * Move dht's validate=True to P2P?
          * In future, can we merge hivemind.DHT into libp2p's DHT? (expiration, conflict resolution, atomic subkey addition, performance)

* __Convert `hivemind.dht.DHT`__
   * [x] support the equivalent of gRPC.unary_unary RPC - get one message, return one message
   * [x] hivemind.dht.DHTProtocol to p2pd
   * [ ] ensure that we can deal with stale connections
   * [x] make sure it passes `tests/test_dht_node.py`
   * [x] make sure it passes `tests/test_dht_experts.py`
   * [ ] tune performance for `tests/benchmark_dht` with 1000 nodes and 16384 experts

* __Convert `hivemind.client.averaging.DecentralizedAverager` to p2pd__
   * [ ] figure out how to implement rpc_join_group:
       * maybe support the equivalent of gRPC.stream_stream RPC (for averager)
   * [ ] ensure that large messages do not cause reduced throughput
       * [ ] test if split_into_parts is still required. If not, remove it
   * [ ] make sure it passes `tests/test_averaging.py`
   * [ ] tune performance with `tests/benchmark_averaging.py`

* __Convert `hivemind.dht.Server`__
   * [ ] find some way to attach several processes to one RPC (as in server/connection_handler.py)
   * [ ] make sure it passes `tests/test_moe.py`
   * [ ] make sure it passes `tests/test_training.py`
   * [ ] tune performance in `tests/benchmark_througphput.py`

* __`Train something from behind NAT`__
    * [ ] make sure ALBERT-v2-large runs
    * [ ] ... with at least one participant under NAT
    * [ ] ... with all but one participant under NAT
    * [ ] ... for at least 48 hours

* Bonus quests:
    * [ ] investigate: __try libp2p relays!__ (TURN)

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -28,6 +45,38 @@ def proto_compile(output_path):
...
+    # check go version:
+    try:
+        proc = subprocess.Popen(['go', 'version'],
+                                stdout=subprocess.PIPE)
+        result, _ = proc.communicate()
+        result = result.decode('ascii', 'replace')
+        _, _, version, _ = result.split(' ')
+        version = version.lstrip('go')
+    
+        if version < "1.13":
```
Comment (justheuristic) [inline setup.py:58]: TODO safe version comparison

Comment (justheuristic) [reply setup.py:58]: tuple comparison or packaging.version

File: setup.py [hunk 2]
```diff
@@ -28,6 +45,38 @@ def proto_compile(output_path):
...
+        urllib.request.urlretrieve(url, os.path.join(tempdir, dest))
+            
+        tar = tarfile.open(dest, 'r:gz')
+        tar.extractall(tempdir)
+        tar.close()
+            
+        with cd(os.path.join(tempdir, 'go-libp2p-daemon-master')):
+            status = os.system('go install ./...')
+            if status:
+                raise RuntimeError('Failed to build or install libp2p-daemon:'\
```
Comment (justheuristic) [inline setup.py:76]: make sure we install p2pd inside hivemind folder

File: setup.py [hunk 3]
```diff
@@ -40,6 +89,11 @@ def run(self):
         super().run()
 
 
+class LibP2PInstall(install):
+    def run(self):
+        install_libp2p_daemon()
+
+
```
Comment (justheuristic) [inline setup.py:96]: maybe make a CI job that automatically saves compiled p2pd

File: setup.py [hunk 4]
```diff
@@ -28,6 +45,38 @@ def proto_compile(output_path):
             file.truncate()
 
 
+def install_libp2p_daemon():
+    # check go version:
+    try:
+        proc = subprocess.Popen(['go', 'version'],
+                                stdout=subprocess.PIPE)
```
Comment (mryab) [inline setup.py:52]: TODO find a popular/frequently used way to package and install Go dependencies for Python projects

File: setup.py [hunk 5]
```diff
@@ -28,6 +46,38 @@ def proto_compile(output_path):
...
+    
+        if version.parse(v) < version.parse("1.13"):
+            raise EnvironmentError(f'newer version of go required: must be >= 1.13, found {version}')
+
+    except FileNotFoundError:
+        raise FileNotFoundError('could not find golang installation')
+
+    with tempfile.TemporaryDirectory() as tempdir:
+        url = 'https://github.com/libp2p/go-libp2p-daemon/archive/master.tar.gz'
+        dest = os.path.join(tempdir, 'libp2p-daemin.tar.gz')
```
Comment (justheuristic) [inline setup.py:67]:
```suggestion
        dest = os.path.join(tempdir, 'libp2p-daemon.tar.gz')   
```

**Comment by borzunov** [comment]
The code from this PR has been finished and merged in #238, #296, and other smaller PRs.

The checklist has been moved to the issue #314.

**Pull request closed without merging**